fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,809 @@
+        setenv_lock(&name, &value, true);
+    } else {
+        unsetenv_lock(&name);
+    }
+
+    extern "C" {
+        fn tzset();
+    }
+
+    unsafe {
+        tzset();
+    }
+}
+
+/// Update the value of [`FISH_EMOJI_WIDTH`].
+fn guess_emoji_width(vars: &dyn Environment) {
+    use crate::fallback::FISH_EMOJI_WIDTH;
+
+    if let Some(width_str) = vars.get(L!("fish_emoji_width")) {
+        // The only valid values are 1 or 2; we default to 1 if it was an invalid int.
+        let new_width = fish_wcstoi(width_str.as_string().chars()).unwrap_or(1);
+        let new_width = new_width.clamp(1, 2);
+        FISH_EMOJI_WIDTH.store(new_width, Ordering::Relaxed);
+        FLOGF!(
+            term_support,
+            "Overriding default fish_emoji_width w/",
+            new_width
+        );
+        return;
+    }
+
+    let term = vars
+        .get(L!("TERM_PROGRAM"))
+        .map(|v| v.as_string())
+        .unwrap_or_else(WString::new);
+    let version = vars
+        .get(L!("TERM_PROGRAM_VERSION"))
+        .map(|v| v.as_string().to_string())
+        .and_then(|v| v.parse().ok())
+        .unwrap_or(0);
+
+    match term.to_string().as_str() {
+        "Apple_Terminal" if version >= 400 => {
+            // Apple Terminal on High Sierra
+            FISH_EMOJI_WIDTH.store(2, Ordering::Relaxed);
+            FLOGF!(term_support, "default emoji width: 2 for", term);
+        }
+        "iTerm.app" => {
+            // iTerm2 now defaults to Unicode 9 sizes for anything after macOS 10.12
+            FISH_EMOJI_WIDTH.store(2, Ordering::Relaxed);
+            FLOGF!(term_support, "default emoji width 2 for iTerm2");
+        }
+        _ => {
+            // Default to whatever the system's wcwidth gives for U+1F603, but only if it's at least
+            // 1 and at most 2.
+            let width = crate::fallback::wcwidth('😃').clamp(1, 2);
+            FISH_EMOJI_WIDTH.store(width, Ordering::Relaxed);
+            FLOGF!(term_support, "default emoji width:", width);
+        }
+    }
+}
+
+/// React to modifying the given variable.
+pub fn env_dispatch_var_change(key: &wstr, vars: &EnvStack) {
+    use once_cell::sync::Lazy;
+
+    // We want to ignore variable changes until the dispatch table is explicitly initialized.
+    if let Some(dispatch_table) = Lazy::get(&VAR_DISPATCH_TABLE) {
+        dispatch_table.dispatch(key, vars);
+    }
+}
+
+fn handle_fish_term_change(vars: &dyn Environment) {
+    update_fish_color_support(vars);
+    crate::ffi::reader_schedule_prompt_repaint();
+}
+
+fn handle_change_ambiguous_width(vars: &dyn Environment) {
+    let new_width = vars
+        .get(L!("fish_ambiguous_width"))
+        .as_ref()
+        .map(|v| v.as_string())
+        // The only valid values are 1 or 2; we default to 1 if it was an invalid int.
+        .map(|fish_ambiguous_width| fish_wcstoi(fish_ambiguous_width.chars()).unwrap_or(0))
+        .unwrap_or(1)
+        .clamp(1, 2);
```

Comment:
The C++ code doesn't clamp this; rather it just sets it to at least zero.

Reply:
Sorry, I confused myself tracking down the different comments in all the headers and cpp files around the g_fish_ambiguous_width and g_fish_emoji_width variables (they were even declared in multiple headers as extern). I'll change that and fix the comment.